### ISSUE TITLE
Accessibility support for SVG output

### ISSUE THREAD

**Issue by fernandoherreradelasheras** [rism-digital/verovio#4273]
It would be great to have accessibility support in Verovio's SVG output. That would make music scores navigable and understandable for users of assistive technologies. Currently, SVG output from Verovio lacks semantic information about musical elements, making it difficult or impossible for assistive technologies to convey meaningful information about the score's content. Accessibility is critical to enable access to digial music content to users with visual impairments.

I have started a quick proof-of-concept implementation that could be useful for kick-starting this discussion  with the community about the best approach:

https://github.com/fernandoherreradelasheras/verovio/tree/accessibility

# Open questions for discussion

## Where should accessibility be implemented

The first main question is: should Verovio implement this accessibility support or should it be a separate SVG DOM Transformation Tool? 

Pros of having a11y inside Verovio:
- Consistent accessibility across all applications using Verovio
- Direct access to musical semantics during rendering that might be more complicated to get
- No post-processing overhead
- Accessibility as a first class citizen means it is always in sync with the latest code

Cons of having a11y inside Verovio:
- More code to maintain in Verovio
- Less flexibility for implementing custom behaviours

Pros of having a11y a separate SVG DOM transformation tool:
- Keeps core library focused
- Different transformation tools could be created for people interested on different models

Cons of having a11y a separate SVG DOM transformation tool:
- Requires parsing and understanding the SVG structure 
- Changes on the SVG structure would break tools
- Duplicates effort across applications

## Scope: Descriptions vs. Descriptions + Navigation

Ideally the SVG produced by Verovio should be ready to be consumed by assistive technologies as it is, without the need of web applications or screen readers specific support. That means providing not only the description of the elements but also the navigation tree.

Generating a ready-to-consume accessibility tree from a navigation perspective is sometimes difficult and, from my experience, particularly complicated for music. Assistive technologies, APIs, and tools are modeled with text document structures in mind, where the main content consists of text and controls distributed at different hierarchical levels. Most assistive tools implement inspection workflows that handle textual cases out of the box with satisfactory results. However, the model for inspecting music can be quite different, as people read scores 
in various ways depending on several factors. A user might want to inspect a score horizontally, moving forward and backward within the same staff, or read the contents of all staves at a given time position, etc.

Generating a ready to be consumed accessibility tree from the navigation point of view is sometimes difficult and, from my experience, in the particular case of music could be extremely complicated. Assistive technologies, APIs and tools are modeled with text document models in mind where the main content are text and controls distributed at different hierarchical levels. Most assistive tools implement inspection workflows that can manage textual cases out of the box with a satisfactory experience. However the model for inspecting music can be quite different as people read scores in different ways depending on several factors. An user might want to inspect an score horizontally moving forward and backward in the same staff or read the contents of all staves at a given time position, etc... 

I think that trying to use ARIA properties and relationships to map the music SVG DOM into a ready-to-use accessible tree for most common music reading use cases is overly ambitious and complicated. Instead of relying on standard assistive tool navigation, web applications displaying the SVG should implement navigation themselves.

I don't have a clear idea yet on how that navigation could be best implemented, and that's another reason why I think it should not be implemented at the SVG level in Verovio, giving different applications space to experiment with different ideas.

## Text Translations

Where should accessible text descriptions live? One option is to use symbolic strings, so Verovio would set as textual descriptions things like  "half-note-b4", "measure-24", "time-signature-3/4" and then any web application should handle the translations. The other option would be to use built-in translations and either generate different svg based on the chosen language or use <swicth> elements with different <g systemLanguage="xx"> elements for each language.

The first option gives applications flexibility in tweaking the presentation but sacrifices readiness-to-use. The latter implies that Verovio would need to maintain translations and potentially increase SVG file size with multiple language variants or multiple SVG variants.

## Community engagement 

We definitely need assistive technology users engaged in this discussion to explore different options and ensure the solutions meet real-world accessibility needs.

# Quick demo

To demonstrate, I have set up a quick test implementing one possible navigation model based on linear inspection of the currently selected staff, allowing users to move forward or backward by elements or measures. Users can also move "up" or "down" to change the currently selected staff. The test, using an SVG generated with my PoC branch, can be seen here:

https://fer.name/verovio-a11y-poc/

**Comment by fernandoherreradelasheras** [comment]
> Would it be better to improve accessibility in the source MEI/MusicXML format, rather than in the SVG output? Could a screen reader traverse the source format directly, without SVG as an intermediary?

From a technical point of view, probably yes. However, universal access advocators discourage having separate versions or custom views targeting people with accessibility requirements. Rationale for this is that separate versions usually become outdated or even abandoned. As they have lower target audience numbers they quickly become low priority.  Forcing the same view for everyone, with accessibility as a first class citizen, if often the best way to ensure that people with special requirements can get still get a satisfactory experience.

**Comment by ahankinson** [comment]
It would probably be most helpful if you could post a minimal example of the kind of SVG you would expect to be produced by Verovio.

**Comment by fernandoherreradelasheras** [comment]
> It would probably be most helpful if you could post a minimal example of the kind of SVG you would expect to be produced by Verovio.

The SVG that the demo is loading is this: 

https://fer.name/verovio-a11y-poc/test_001.svg

Accessibility technologies will describe any object present in the accessibility tree using its textual representation. When the object is a `<text>` element, the textual representation is the text content itself. Otherwise this information would be extracted from `aria-label`/`aria-description` of the element or from `<title>`/`<description>` children elements content.

So basically any element of the svg that that we want to be visible by assistive technologies need to include its description in any of those fields. For example, what is currently represented as:
```
<g id="j15k8r0t" class="clef">
    <use xlink:href="[#E050-zsbuqix](view-source:https://fer.name/verovio-a11y-poc/test_001.svg#E050-zsbuqix)" transform="translate(1636, 960) scale(0.64, 0.64)" />
</g>
```
should be something like:
```
<g id="j15k8r0t" class="clef">
    <title>G clef 2nd line</title>
    <use xlink:href="[#E050-zsbuqix](view-source:https://fer.name/verovio-a11y-poc/test_001.svg#E050-zsbuqix)" transform="translate(1636, 960) scale(0.64, 0.64)" />
</g>
```

In my demo I used `<title>` instead of `<description>`/`aria-description`  because the former produces alto tooltips that are helpful for quick inspection. 

If we don't want to have one element present in the accessible tree we should add the attribute `aria-hidden` like in this example from the demo:
```

                       <g id="h14jxjpl" class="verse">
                           <title>Verse 1 Un</title>
                           <g id="i1ogjlmn" class="syl" aria-hidden="true">
                              <text x="3941" y="1558" font-size="0px">
                                 <tspan id="jlefbsu" class="text">
                                    <tspan font-size="360px">Un</tspan>
                                 </tspan>
                              </text>
                           </g>
                        </g>
```

in this example, as mapping `<verse>` and `<syl>` objects directly into the accessible tree would result in a too verbose/redundant navigation ([next element] -> "verse 1" [next element] -> "syl Un" ) I hide the `<syl>` elements and collected all their contents in the `<verse>` title.

**Comment by ahankinson** [comment]
I guess what I meant is that the actual titles were supplied "by hand" -- That is, you supplied the titles for individual symbols in the source code of your branch, c.f. https://github.com/rism-digital/verovio/compare/develop...fernandoherreradelasheras:verovio:accessibility#diff-7124361364eeb9afde6af8a165c21d610a6639a0ae7e8465dea2a9d685fdb1c4R227-R280

This is different from Verovio auto-generating the titles from the musical symbols themselves, e.g., in your example above, you could expect Verovio to generate `<title>Verse 1 Un</title>` from the fact that there is a `class="verse"`, that it is the first line, and that the syllable given is "Un". (This is just an illustration, of course, but all the information is already there in the SVG to generate that title.) 

The challenge with adding fixed strings is that Verovio is already operating with tight resource constraints, particularly around browser memory and library size for the JavaScript builds. Fixed strings can add up quite quickly since they're not able to be "minimized" like other javascript code, so have to take up the full space. If you can figure out a way to auto-generate the titles from existing musical content, though, it's less likely to have an impact on the size of the binary.
